### ISSUE TITLE
Under high load - TaskCanceledException Thrown

### ISSUE THREAD

**Issue by spsingletary** [object/Simple.OData.Client#305]
While load testing an application with ~4,000 virtual users, the Simple OData Client begins to throw TaskCanceledExceptions which causes the application to fail that request. Why isn't this throwing a more specific error message such as SockException?.

`System.Threading.Tasks.TaskCanceledException:
   at Microsoft.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (Microsoft.Threading.Tasks, Version=1.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Microsoft.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccess (Microsoft.Threading.Tasks, Version=1.0.12.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a)
   at Simple.OData.Client.FluentClientBase`1+<RectifyColumnSelectionAsync>d__4.MoveNext (Simple.OData.Client.Net40, Version=4.23.3.0, Culture=neutral, PublicKeyToken=e33b67d3bb5581e4)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at TECO.CSSP.Portal.Gateway.NetWeaver.ERP.ERPGateway+<>c__DisplayClass12_0+<<GetAccountAddressIndependentEmailsByAccountIdAsync>b__2>d.MoveNext (MyApp)
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at System.Runtime.CompilerServices.TaskAwaiter`1.GetResult (mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089)
   at MyApp+<GetAccountAddressIndependentEmailsByAccountIdAsync>d__12.MoveNext`

Here is the code being instrumented.

`
public async Task<ICollection<AccountAddressIndependentEmail>> GetAccountAddressIndependentEmailsByAccountIdAsync(string accountId)
        {
            try
            {
                var target = new Regex(RegexPatterns.Numeric);
                var isNumber = target.IsMatch(accountId);

                if (!isNumber)
                {
                    throw new ArgumentException("Invalid account id: " + accountId);
                }

                var settings = new ODataClientSettings()
                {
                    BaseUri = this.ServiceUri,
                    BeforeRequest = (x) => BeforeRequest(x, HttpRequestHeaderOptions.RequestedWithXml),
                    OnTrace = (format, args) => OnTrace(format, args),
                    PayloadFormat = ODataPayloadFormat.Json,
                    Credentials = this.credentials,
                    MetadataDocument = await this.GetMetadataDocument()
                };

                var client = new ODataClient(settings);

                var result = await RetryUtility.RetryAction<Task<IEnumerable<AccountAddressIndependentEmail>>>(
                    async () => await client.For<Account>()
                    .Filter(x => x.AccountID == accountId)
                    .NavigateTo<AccountAddressIndependentEmail>()
                    .FindEntriesAsync(),
                    3,
                    1000,
                    this.AppInsightsLogger);

                return result.ToList();
            }
            catch (WebRequestException ex)
            {
                var error = new NetWeaverError(ex.Response);
                AppInsightsLogger.Error(ex, "Exception: " + error.Message);
            }
            catch (Exception ex)
            {
                AppInsightsLogger.Error(ex, "Exception: " + ex.Message);
            }

            return null;
        }`

**Comment by sixlettervariables** [comment]
Have you inspected the InnerException? The TaskCancelledException is probably being raised from within the underlying HttpClient.

**Comment by object** [comment]
Yes, as noted above, you should inspect the InnerException of TaskCancelledException.

**Comment by spsingletary** [comment]
Based on the responses provided, I added additional logic to the catch block to throw the inner exception if it exists. After performing more testing, no inner exception was thrown, but the original TaskCanceledException was thrown.

**Comment by object** [comment]
Simple.OData.Client uses internally HttpClient that must be throwing such exception. Can you check how long time does it take to execute requests under high load? One possible reason is that HttpClient encounters timeout while waiting for the response - if no response is received from the server within X seconds it will cancel the task.
